rinja/src/filters/builtin.rs

```diff
@@ -528,6 +528,215 @@ pub fn title(s: impl fmt::Display) -> Result<String, fmt::Error> {
     Ok(output)
 }
 
+/// For a value of `±1` by default an empty string `""` is returned, otherwise `"s"`.
+///
+/// # Examples
+///
+/// ## With default arguments
+///
+/// ```
+/// # #[cfg(feature = "code-in-doc")] {
+/// # use rinja::Template;
+/// /// ```jinja
+/// /// I have {{dogs}} dog{{dogs|pluralize}} and {{cats}} cat{{cats|pluralize}}.
+/// /// ```
+/// #[derive(Template)]
+/// #[template(ext = "html", in_doc = true)]
+/// struct Pets {
+///     dogs: i8,
+///     cats: i8,
+/// }
+///
+/// assert_eq!(
+///     Pets { dogs: 0, cats: 0 }.to_string(),
+///     "I have 0 dogs and 0 cats."
+/// );
+/// assert_eq!(
+///     Pets { dogs: 1, cats: 1 }.to_string(),
+///     "I have 1 dog and 1 cat."
+/// );
+/// assert_eq!(
+///     Pets { dogs: -1, cats: 99 }.to_string(),
+///     "I have -1 dog and 99 cats."
+/// );
+/// # }
+/// ```
+///
+/// ## Overriding the singular case
+///
+/// ```
+/// # #[cfg(feature = "code-in-doc")] {
+/// # use rinja::Template;
+/// /// ```jinja
+/// /// I have {{dogs}} dog{{ dogs|pluralize("go") }}.
+/// /// ```
+/// #[derive(Template)]
+/// #[template(ext = "html", in_doc = true)]
+/// struct Dog {
+///     dogs: i8,
+/// }
+///
+/// assert_eq!(
+///     Dog { dogs: 0 }.to_string(),
+///     "I have 0 dogs."
+/// );
+/// assert_eq!(
+///     Dog { dogs: 1 }.to_string(),
+///     "I have 1 doggo."
+/// );
+/// # }
+/// ```
+///
+/// ## Overriding singular and plural cases
+///
+/// ```
+/// # #[cfg(feature = "code-in-doc")] {
+/// # use rinja::Template;
+/// /// ```jinja
+/// /// I have {{mice}} {{ mice|pluralize("mouse", "mice") }}.
+/// /// ```
+/// #[derive(Template)]
+/// #[template(ext = "html", in_doc = true)]
+/// struct Mice {
+///     mice: i8,
+/// }
+///
+/// assert_eq!(
+///     Mice { mice: 42 }.to_string(),
+///     "I have 42 mice."
+/// );
+/// assert_eq!(
+///     Mice { mice: 1 }.to_string(),
+///     "I have 1 mouse."
+/// );
+/// # }
+/// ```
+#[inline]
+pub fn pluralize<C, S, P>(count: C, singular: S, plural: P) -> Result<Pluralize<S, P>, C::Error>
+where
+    C: PluralizeCount,
+{
+    match count.is_singular()? {
+        true => Ok(Pluralize::Singular(singular)),
+        false => Ok(Pluralize::Plural(plural)),
+    }
+}
+
+/// An integer that can have the value `+1` and maybe `-1`.
+pub trait PluralizeCount {
+    /// A possible error that can occur while checking the value.
+    type Error: Into<Error>;
+
+    /// Returns `true` if and only if the value is `±1`.
+    fn is_singular(&self) -> Result<bool, Self::Error>;
```

Comment:
Is there an implementation which can return something else than `Ok`?

Reply:
Because I (intentionally) didn't seal the trait, I want to keep it as generic as possible. If you wanted to, you could implement a fallible lookup for your custom type. E.g. acquiring a lock or reading from a database. I would not advise to do that in a template, but it does not hurt make it possible. :)